File: cpp/manual_p2p.cpp
#include "p2p.hpp"
#include <iostream>
#include <string>
#include <time.h>
#include <chrono>
#include <map>

using namespace std;
using namespace p2p;
using namespace socket_tool;

inline chrono::system_clock::time_point time_now(){
    return chrono::system_clock::now();
}

inline int duration_ms( chrono::system_clock::time_point t1, chrono::system_clock::time_point t2 ){
    return ( chrono::duration<double>(t2 - t1).count() * 1000 );
}

int main(){

    srand(time(NULL));
    int my_port = 50000 + (rand()%1000);
    cout << "listening on 0.0.0.0:" << my_port << "\n";

    string target_ip;
    int target_port;

    cout << "enter ip:  ";
    cin >> target_ip;
    cout << "enter port:";
    cin >> target_port;

    try{

        P2P_connection connection{my_port, Address{target_ip, target_port}};

        connection.punch_hole();

        chrono::system_clock::time_point time_point_table[100] = {};
        int echo_num;

        char buff[1201]={0};

        size_t idx = 0;

        while (1){ try{

            int random_num = rand() % 100;
            time_point_table[random_num] = time_now();
            size_t n = snprintf( buff, 1200, "%d %d", random_num, echo_num);
            

            connection.send( buff, n );

            n=1200;
            connection.recv( buff, &n );
            buff[n] = '\0';
            sscanf( buff, "%d %d", &echo_num, &random_num);

            cout << ++idx << ".: key num: " << random_num << "\t echo num: " << echo_num << " \t\t round tripp time: " << duration_ms( time_now(), time_point_table[random_num%100] ) << "\tus\n";

        }catch(Timeout_excpetion _){
            connection.punch_hole();
        }catch(Exception e){
            cerr << "cauth exception " << e.message << "\n";
        }
        }


    }catch( Exception e ){
        cerr << "uncouth exception " << e.message << "\n";
    }catch( Error e ){
        cerr << "error " << e.message << "\n";
    }catch(...){
        cerr << "unknown \n";
    }


}

File: python/manual_p2p.py
import socket
import struct
import p2p
import os
import random
import time
import json

STUN_SERVER = ("stun.l.google.com", 19302)
MAGIC_COOKIE = 0x2112A442


def build_binding_request():
    # STUN Binding Request
    msg_type = 0x0001
    msg_length = 0
    transaction_id = os.urandom(12)

    packet = struct.pack(
        "!HHI12s",
        msg_type,
        msg_length,
        MAGIC_COOKIE,
        transaction_id,
    )

    return packet, transaction_id


def parse_response(data, expected_transaction_id):
    if len(data) < 20:
        raise ValueError("Response too short")

    msg_type, msg_length, magic_cookie = struct.unpack("!HHI", data[:8])
    transaction_id = data[8:20]

    if msg_type != 0x0101:
        raise ValueError(f"Unexpected message type: {hex(msg_type)}")

    if magic_cookie != MAGIC_COOKIE:
        raise ValueError("Invalid magic cookie")

    if transaction_id != expected_transaction_id:
        raise ValueError("Transaction ID mismatch")

    offset = 20

    while offset + 4 <= len(data):
        attr_type, attr_length = struct.unpack(
            "!HH", data[offset:offset + 4]
        )

        value = data[offset + 4:offset + 4 + attr_length]

        if attr_type == 0x0020:  # XOR-MAPPED-ADDRESS

            if value[1] == 1:  # IPv4

                xor_port = struct.unpack("!H", value[2:4])[0]
                port = xor_port ^ (MAGIC_COOKIE >> 16)

                xor_ip = struct.unpack("!I", value[4:8])[0]
                ip = xor_ip ^ MAGIC_COOKIE

                ip = socket.inet_ntoa(struct.pack("!I", ip))

                return ip, port

            elif value[1] == 2:  # IPv6
                raise NotImplementedError("IPv6 not implemented")

        # Attributes are padded to a multiple of 4 bytes
        offset += 4 + ((attr_length + 3) & ~3)

    raise ValueError("No XOR-MAPPED-ADDRESS attribute found")




if __name__ == "__main__":
    sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    sock.bind(("0.0.0.0", 0))
    sock.settimeout(5)
    
    request, transaction_id = build_binding_request()
    
    sock.sendto(request, STUN_SERVER)
    
    response, addr = sock.recvfrom(2048)
    
    print("Received response from", addr)
    
    ip, port = parse_response(response, transaction_id)

    print( f"listening on {ip}:{port}" )
    target = input( "enter target (x.x.x.x:port) -& " ).split(':')

    connection = p2p.P2PConnection( sock, target[0], int(target[1]) )

    connection.punch_hole()


    nums = {}
    send_back = []
    while 1:
        try:
            num = random.randrange(1, 100000)
            nums[num] = time.time()
            
            data = json.dumps({
                "echo": send_back,
                "send": num
            }).encode("utf-8")
            
            connection.send(data)
            data = connection.recv()
            
            data = json.loads(data)
            send_back = []
            send_back.append(data["send"])
            
            for echoed_num in data["echo"]:
                print( f"{round((time.time() - nums[echoed_num])*1000, 3)} ms -------> {echoed_num}" )
                del nums[echoed_num]

        except socket.timeout:
            connection.punch_hole()
    
    sock.close()
    